src/shogun/mathematics/linalg/Vector.cpp

```diff
@@ -0,0 +1,226 @@
+}
+
+template<class T>
+Vector<T>::Vector(Vector<T> const &vector)
+{
+	init();
+
+	m_data = std::shared_ptr<T>(reinterpret_cast<T*>(SG_MALLOC(aligned_t, vector.size())),
+									[](T* ptr) {SG_FREE(ptr);});
+	m_len = vector.m_len;
+
+
+	if (vector.onGPU())
+	{
+#ifdef HAVE_VIENNACL
+		m_onGPU = true;
+		m_gpu_impl = std::unique_ptr<GPUVectorImpl>(new GPUVectorImpl(*(vector.m_gpu_impl)));
+		transferToCPU();
+#endif
+	}
+	else
+	{
+		std::copy(vector.data(), vector.data() + vector.size(), m_data.get());
+	}
+}
+
+template<class T>
+Vector<T>::~Vector()
+{
+}
+
+template<class T>
+Vector<T>& Vector<T>::operator=(SGVector<T> const &vector)
+{
+	m_data = std::shared_ptr<T>(reinterpret_cast<T*>(SG_MALLOC(aligned_t, vector.vlen)),
+									[](T* ptr) {SG_FREE(ptr);});
+	std::copy(vector.vector, vector.vector+vector.vlen, m_data.get());
+	m_len = vector.vlen;
+	m_onGPU = false;
+	m_gpu_impl.release();
+	return *this;
+}
+
+template<class T>
+Vector<T>& Vector<T>::operator=(Vector<T> const &vector)
+{
+	m_data = std::shared_ptr<T>(reinterpret_cast<T*>(SG_MALLOC(aligned_t, vector.size())),
+	 							[](T* ptr) {SG_FREE(ptr);});
+	m_len = vector.m_len;
+
+	if (vector.onGPU())
+	{
+#ifdef HAVE_VIENNACL
+		m_onGPU = true;
+		m_gpu_impl.reset(new GPUVectorImpl(*(vector.m_gpu_impl)));
+		transferToCPU();
+#endif
+	}
+	else
+	{
+		m_onGPU = false;
+		std::copy(vector.data(), vector.data() + vector.size(), m_data.get());
+	}
+	return *this;
+}
+
+template<class T>
+Vector<T>::operator SGVector<T>()
+{
+	if (onGPU())
+		transferToCPU();
+	SGVector<T> vector(m_len);
+	std::copy(m_data.get(), m_data.get()+m_len, vector.vector);
+	return vector;
+}
+
+template<class T>
+bool Vector<T>::onGPU() const
+{
+	return m_onGPU;
+}
+
+template<class T>
+T* Vector<T>::data()
+{
+	if (onGPU())
+		transferToCPU();
+	return m_data.get();
+}
+
+template<class T>
+T const * Vector<T>::data() const
+{
+	return m_data.get();
+}
+
+template<class T>
+index_t Vector<T>::size() const
+{
+	return m_len; // Assume GPU operations won't change the length
+}
+
+template<class T>
+void Vector<T>::transferToGPU()
+{
+#ifdef HAVE_VIENNACL
+	m_gpu_impl = std::unique_ptr<GPUVectorImpl>(new GPUVectorImpl(m_data.get(), m_len));
+	m_onGPU = true;
+#else
+	SG_SINFO("Transfer incomplete. No registered GPU backend found.\n");
+#endif
+}
+
+template<class T>
+void Vector<T>::releaseFromGPU()
+{
+#ifdef HAVE_VIENNACL
+	m_gpu_impl.reset(nullptr);
+	m_onGPU = false;
+#endif
+}
+
+template<class T>
+void Vector<T>::transferToCPU()
+{
+	if (!onGPU())
+	{
+		SG_SINFO("There is no data to transfer on GPU.\n")
+		return;
+	}
+
+#ifdef HAVE_VIENNACL
+	m_gpu_impl->transferToCPU(m_data.get()); // Assume the transfer won't change the length of the vector
+#endif
+}
+
+template<class T>
+T& Vector<T>::operator[](index_t index)
+{
+// Doesn't seem to work
+/*
+#ifdef HAVE_VIENNACL
+	if (onGPU())
+		return (*m_gpu_impl)[index];
+#endif
+*/
+	if (onGPU())
+		transferToCPU();
+	return m_data.get()[index];
```

Comment:
I was using (*m_gpu_impl)[index] to access GPU vector element in unit-tests.
But I made some changes here. Ideally, I want [] to return scalar type from GPU when`onGPU()==true` and from CPU when `onGPU()==false`.
So, the `m_gpu_impl` can stay private while the unit-test can still access the element.

However,
- if I do it as the way I commented out, apparently the two types cannot be converted easily.
- if I do `transferToCPU()`, a - it's extremely slow; b - i cannot use it for `operator[](index_t index) const`; c - I cannot change the value of the element (may not required by users? I was changing the value in unit-tests).

So I don't have very good idea to implement the method here.